### ISSUE TITLE
Delay observed in netty to read a request from a channel

### ISSUE THREAD

**Issue by otterc** [netty/netty#9890]
### Summary 
We run Spark 2.3 on Yarn. In Spark, we use netty as the network communication framework. 
We have been facing issues with the netty server in our Spark External Shuffle service. During peak load, SASL requests submitted by the clients to this service timeout after 30 seconds. 

We investigated the SASL timeouts from different angles. We took a tcp dump and could see that there is no network delay or retransmissions. The tcp dump shows that the request was received by the server within milliseconds but the server sends a payload back after 30 seconds.
We have observed that the netty server takes more than 30 seconds to read a message from a channel and after it is read, it is processed in a millisecond. This leads us to believe that the issue maybe caused by something in netty. More debugging details are provided below.

### Details
There are 2 types of requests handled by our shuffle service:
1. Requests whose handlers execute quickly. They are handled by the default Server IO event loop group. **SASL** requests are such requests with a timeout of 30 seconds. 
2. Requests that do random disk IO.  These requests have no timeout and there is a dedicated event loop group to handle these requests. 

The netty pipeline looks like:
```
channel.pipeline()
    .addLast("encoder", ENCODER)
    .addLast(TransportFrameDecoder.HANDLER_NAME, NettyUtils.createFrameDecoder())
    .addLast("decoder", DECODER)
    .addLast("idleStateHandler",
        new IdleStateHandler(0, 0, conf.connectionTimeoutMs() / 1000))
    // NOTE: Chunks are currently guaranteed to be returned in the order of request, but this
    // would require more logic to guarantee if this were not part of the same event loop.
    .addLast("handler", channelHandler)
    // Use a separate EventLoopGroup to handle ChunkFetchRequest messages.
    .addLast(chunkFetchWorkers, "chunkFetchHandler", chunkFetchHandler);
```

Sizes of event loop groups:
- Server IO event loop group:  80
- Chunk fetch event loop group: 45 

#### Handler Analysis
The handling of SASL requests itself takes 0 - 1 milliseconds.
I added extra handlers in netty pipeline to record the time when a request is read and when its response is flushed to check if the Decoder/Encoder is taking any time.
```
channel.pipeline()
    .addLast("incoming", new FirstHandler())
    .addLast("outgoing", new LastHandler())
    .addLast("encoder", ENCODER)
    .addLast(TransportFrameDecoder.HANDLER_NAME, NettyUtils.createFrameDecoder())
    .addLast("decoder", DECODER)
    .addLast("idleStateHandler",
        new IdleStateHandler(0, 0, conf.connectionTimeoutMs() / 1000))
    .addLast("handler", channelHandler)
    // Use a separate EventLoopGroup to handle ChunkFetchRequest messages.
    .addLast(chunkFetchWorkers, "chunkFetchHandler", chunkFetchHandler);
```
Implementation of FirstHandler ->
```
@Override
public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
  if (!ctx.channel().hasAttr(INITIAL_REQUEST_RECV)) {
    ctx.channel().attr(INITIAL_REQUEST_RECV).set(System.currentTimeMillis());
  }
  super.channelRead(ctx, msg);
}
```

It was observed that the request itself was read 30 seconds after client had sent it.
Client log :
```
2019-12-13 20:33:43,609 TRACE org.apache.spark.network.client.TransportClient: Sending request 8863541504479811769 to 10.150.16.27:7337 took 17 ms
```
Server log :
```
2019-12-13 20:34:13,895 INFO org.apache.spark.network.server.LastHandler OUTPUT request 8863541504479811769 channel 6fc450d5 request_rec 1576269253894 flush 1576269253895
```
**The client sent the request at 20:33:43 but the server starts reading the request at 1576269253894 which is 20:34:13**.

To find out why there is a delay to read the request, I looked at the NioEventLoop.run() method. Specifically, this part
```
try {
    processSelectedKeys();
} finally {
    // Ensure we always run tasks.
    final long ioTime = System.nanoTime() - ioStartTime;
    runAllTasks(ioTime * (100 - ioRatio) / ioRatio);
}
```
I recorded the time it takes to run `processSelectedKeys` and `runAllTasks` for a single event loop. 
The `taskQueue` of a ServerIO event loop contains `WriteAndFlushTask`s which are added by the chunk fetch event loop.  The ServerIO thread spends more time in `runAllTasks()`, however a single execution of this method takes less than a second.

In the 30 seconds before a request is timed-out, I observed multiple invocations of `processSelectedKeys` and `runAllTasks`.
```
2019-12-13 20:34:12,640 INFO org.spark_project.io.netty.channel.nio.NioEventLoop: shuffle-server-4-1 selectedKeys nanos 4590296 millis 4
2019-12-13 20:34:13,081 INFO org.spark_project.io.netty.channel.nio.NioEventLoop: shuffle-server-4-1 runAllTasks millis 441
2019-12-13 20:34:13,087 INFO org.spark_project.io.netty.channel.nio.NioEventLoop: shuffle-server-4-1 selectedKeys nanos 5735343 millis 5
2019-12-13 20:34:13,149 INFO org.spark_project.io.netty.channel.nio.NioEventLoop: shuffle-server-4-1 runAllTasks millis 62
2019-12-13 20:34:13,150 INFO org.spark_project.io.netty.channel.nio.NioEventLoop: shuffle-server-4-1 selectedKeys nanos 1332499 millis 1
2019-12-13 20:34:13,151 INFO org.spark_project.io.netty.channel.nio.NioEventLoop: shuffle-server-4-1 runAllTasks millis 1
```
**Since, `processSelectedKeys` is called multiple times before the request is eventually read, it is evident that `runAllTasks` is not consuming all the time of the  IOThread.**

#### TCP Dump
From the tcp dump, it seems that a request which times out reaches the server in time. However, there is no response from the server for 30 seconds. This is when the client times out. Here is a snapshot of one such request.
![image](https://user-images.githubusercontent.com/4690923/70932173-eea93c00-1fed-11ea-8ab4-b1937a56c9ca.png)

#### Miscellaneous

- GC 
GC takes sub-seconds. Below are the GC logs during one such instance of SASL timeout.
```
2019-11-12T21:47:17.609+0000: 1623.446: [GC (Allocation Failure) [PSYoungGen: 20096K->1280K(19968K)] 110105K->91449K(542720K), 0.0026238 secs] [Times: user=0.02 sys=0.00, real=0.00 secs]
2019-11-12T21:47:22.759+0000: 1628.595: [GC (Allocation Failure) [PSYoungGen: 19712K->1136K(19456K)] 109881K->91473K(542208K), 0.0026039 secs] [Times: user=0.02 sys=0.00, real=0.00 secs]
2019-11-12T21:47:31.760+0000: 1637.596: [GC (Allocation Failure) [PSYoungGen: 19056K->1152K(18944K)] 109393K->91649K(541696K), 0.0026084 secs] [Times: user=0.03 sys=0.00, real=0.00 secs]
2019-11-12T21:47:38.458+0000: 1644.294: [GC (Allocation Failure) [PSYoungGen: 18560K->1248K(18432K)] 109057K->91913K(541184K), 0.0024125 secs] [Times: user=0.02 sys=0.00, real=0.01 secs]
2019-11-12T21:47:45.772+0000: 1651.608: [GC (Allocation Failure) [PSYoungGen: 18144K->1312K(17920K)] 108809K->92137K(540672K), 0.0026104 secs] [Times: user=0.02 sys=0.00, real=0.00 secs]
2019-11-12T21:47:51.531+0000: 1657.367: [GC (Allocation Failure) [PSYoungGen: 17696K->1376K(17408K)] 108521K->92361K(540160K), 0.0024784 secs] [Times: user=0.02 sys=0.00, real=0.01 secs]
2019-11-12T21:47:57.925+0000: 1663.761: [GC (Allocation Failure) [PSYoungGen: 17248K->1312K(16896K)] 108233K->92449K(539648K), 0.0026063 secs] [Times: user=0.03 sys=0.00, real=0.00 secs]
```

- Network Bandwidth
I have also analyzed the network bandwidth when this happens and the peak sent/receive rate are much less than the max. 

### Expected behavior
Once a channel is ready with a message, we expect netty to start reading the message immediately. 

### Actual behavior
Observed that the messages are being read at a delay of 30 seconds.

### Netty version
4.1.17.Final. I have upgraded the version to 4.1.43.Final as well and observed similar behavior.

### JVM version (e.g. `java -version`)
1.8.0_172

### OS version (e.g. `uname -a`)
Linux  3.10.0-862.11.6.el7.x86_64 #1 SMP Fri Aug 10 16:55:11 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

**Comment by normanmaurer** [comment]
@otterc I wonder if you could add a `LoggingHandler` as the first handler to your pipeline and verify that the bytes are not seen by it when you hit the timeout.

**Comment by otterc** [comment]
@normanmaurer thanks for the suggestion. I will try that.

**Comment by johnou** [comment]
You should also try taking thread dumps, sounds like you may have some blocking IO operations on Netty event loop threads.

**Comment by otterc** [comment]
@normanmaurer Added the `LoggingHandler` as the first handler in the pipeline.
Looks like even the channel `REGISTERED` and `ACTIVE` events are recorded after 30 seconds. 
Client log:
```
2020-01-08 20:25:27,918 TRACE org.apache.spark.network.client.TransportClient: Sending request 8257222101189151324 to 10.150.16.27:7337 took 7 ms
```
Server log:
```
2020-01-08 20:25:58,991 DEBUG org.spark_project.io.netty.handler.logging.LoggingHandler: [id: 0x78bdafea, L:/10.150.16.27:7337 - R:/10.150.16.44:58348] REGISTERED
2020-01-08 20:25:58,991 DEBUG org.spark_project.io.netty.handler.logging.LoggingHandler: [id: 0x78bdafea, L:/10.150.16.27:7337 - R:/10.150.16.44:58348] ACTIVE

2020-01-08 20:25:58,993 DEBUG org.spark_project.io.netty.handler.logging.LoggingHandler: [id: 0x78bdafea, L:/10.150.16.27:7337 - R:/10.150.16.44:58348] READ: 48B
2020-01-08 20:25:58,993 DEBUG org.spark_project.io.netty.handler.logging.LoggingHandler: [id: 0x78bdafea, L:/10.150.16.27:7337 - R:/10.150.16.44:58348] WRITE: org.apache.spark.network.protocol.MessageWithHeader@7dbf60af
2020-01-08 20:25:58,993 DEBUG org.spark_project.io.netty.handler.logging.LoggingHandler: [id: 0x78bdafea, L:/10.150.16.27:7337 - R:/10.150.16.44:58348] FLUSH
2020-01-08 20:25:58,993 DEBUG org.spark_project.io.netty.handler.logging.LoggingHandler: [id: 0x78bdafea, L:/10.150.16.27:7337 - R:/10.150.16.44:58348] READ COMPLETE
```

**Comment by otterc** [comment]
@johnou We are not blocking the default Server IO netty event loop group. In the handler analysis (problem description), the handling of the requests within a millisecond.

**Comment by johnou** [comment]
@otterc did you try taking multiple thread dumps when you are seeing the timeouts? could be something like the SASL handshake blocking the event loop or maybe something like the server has run out of entropy.

**Comment by otterc** [comment]
@johnou I didn't take thread dumps yet. After looking at the result of `LoggingHandler`, I realized that even the channel registration is delayed by 30 seconds. In our Spark External Shuffle service, the boss event group and the worker event group are same. 
```
    EventLoopGroup bossGroup =
      NettyUtils.createEventLoop(ioMode, conf.serverThreads(), conf.getModuleName() + "-server");
    EventLoopGroup workerGroup = bossGroup;

    bootstrap = new ServerBootstrap()
      .group(bossGroup, workerGroup)
      .channel(NettyUtils.getServerChannelClass(ioMode))
      .option(ChannelOption.ALLOCATOR, allocator)
      .childOption(ChannelOption.ALLOCATOR, allocator);
```
Since the boss thread is also the worker thread, channel registration may get delayed?

I ran another experiment with a dedicated event loop group (with 1 thread) as boss group. 
```
    EventLoopGroup bossGroup = NettyUtils.createEventLoop(ioMode, 1,
      conf.getModuleName() + "-boss");
    EventLoopGroup workerGroup =  NettyUtils.createEventLoop(ioMode, conf.serverThreads(),
    conf.getModuleName() + "-server");

    bootstrap = new ServerBootstrap()
      .group(bossGroup, workerGroup)
      .channel(NettyUtils.getServerChannelClass(ioMode))
      .option(ChannelOption.ALLOCATOR, allocator)
      .childOption(ChannelOption.ALLOCATOR, allocator);
```
In this run, I don't see SASL requests timing out.

**Comment by normanmaurer** [comment]
@otterc it should one be delayed if the `EventLoop` is too busy. Whatever this means depends.. While you may not block you may run "cpu intense" workload on it and so it will not be able to accept and register new connections in time. I would do what @johnou said and take multiple thread-dumps + use a profiler to see what is going on.

**Comment by normanmaurer** [comment]
Closing this now... @otterc please reopen if you still thing this is a netty bug.
